File: SourceCode/Python/ChalmersGU/Other/Huffman.py

from functools import total_ordering
from MinHeap import MinHeap

#/* *** ODSATag: HuffTree *** */
@total_ordering
class HuffTree:
    def __init__(self, elem=None, left=None, right=None, weight=None):
        self.elem = elem
        self.left = left
        self.right = right
        self.weight = weight

    def is_leaf(self):
        return self.left is None or self.right is None

    def __lt__(self, other):
        return self.weight < other.weight

    def __eq__(self, other):
        return self.weight == other.weight
#/* *** ODSAendTag: HuffTree *** */

    def encoding_map(self):
        if self.is_leaf():
            return {self.elem: ""}

        result = {}
        left_encoding = self.left.encoding_map()
        for ch, bits in left_encoding.items():
            result[ch] = "0" + bits
        right_encoding = self.right.encoding_map()
        for ch, bits in right_encoding.items():
            result[ch] = "1" + bits
        return result

    def __str__(self):
        enc = self.encoding_map()
        s = "\n".join(f"  {ch}: <{bits}>" for ch, bits in enc.items())
        return f"{self.weight}\n" + s

#/* *** ODSATag: HuffTreeBuild *** */
def build_huff_tree(frequencies):
    huff_heap = MinHeap()
    # Initialise the heap with singleton Huffman trees
    for ch, freq in frequencies.items():
        huff_heap.add(HuffTree(elem=ch, weight=freq))

    # While at least two trees left on heap
    while huff_heap.size() > 1:
        t1 = huff_heap.removeMin()
        t2 = huff_heap.removeMin()
        t3 = HuffTree(left=t1, right=t2, weight=t1.weight + t2.weight)
        # Return new tree to the heap
        huff_heap.add(t3)

    # Return the final Huffman tree
    return huff_heap.removeMin() 
#/* *** ODSAendTag: HuffTreeBuild *** */

if __name__=='__main__':
    freqs = {
        'a': 5,
        'b': 2,
        'c': 1,
        'd': 1,
        'r': 2,
        '!': 1,
    }
    tree = build_huff_tree(freqs)
    print(tree)

File: SourceCode/Java_Generic/ChalmersGU/Other/Huffman.java

/* *** ODSATag: HuffTree *** */
class HuffTree implements Comparable<HuffTree> {
    public char element;
    public int weight;
    public HuffTree left;
    public HuffTree right;

    HuffTree(char el, int wt) {
        element = el; weight = wt;
    }

    HuffTree(HuffTree l, HuffTree r, int wt) {
        left = l; right = r; weight = wt;
    }

    public boolean isLeaf() {
        return left == null || right == null;
    }

    public int compareTo(HuffTree other) {
        return Integer.compare(this.weight, other.weight);
    }
/* *** ODSAendTag: HuffTree *** */

    public Map<Character, String> encodingMap() {
        Map<Character, String> result = new SeparateChainingHashMap<>();
        if (isLeaf()) {
            result.put(element, "");
            return result;
        }

        Map<Character, String> leftEncoding = left.encodingMap();
        for (char ch : leftEncoding)
            result.put(ch, "0" + leftEncoding.get(ch));
        Map<Character, String> rightEncoding = right.encodingMap();
        for (char ch : rightEncoding)
            result.put(ch, "1" + rightEncoding.get(ch));
        return result;
    }

    public String toString() {
        Map<Character, String> enc = encodingMap();
        String s = weight + "\n";
        for (char c : enc) {
            s += "  " + c + ": <" + enc.get(c) + ">\n";
        }
        return s;
    }
/* *** ODSATag: HuffTree *** */
}
/* *** ODSAendTag: HuffTree *** */

class Huffman {
/* *** ODSATag: HuffTreeBuild *** */
public static HuffTree buildHuffTree(Map<Character, Integer> frequencies) {
    MinHeap<HuffTree> huffHeap = new MinHeap<>();
    for (char ch : frequencies) { // Initialise the heap with singleton Huffman trees
        int freq = frequencies.get(ch);
        huffHeap.add(new HuffTree(ch, freq));
    }
    while (huffHeap.size() > 1) { // While at least two trees left on heap
        HuffTree t1 = huffHeap.removeMin();
        HuffTree t2 = huffHeap.removeMin();
        HuffTree t3 = new HuffTree(t1, t2, t1.weight + t2.weight);
        huffHeap.add(t3);         // Return new tree to heap
    }
    return huffHeap.removeMin();  // Return the final Huffman tree
}
/* *** ODSAendTag: HuffTreeBuild *** */

public static void main(String[] args) {
    Map<Character, Integer> freqs = new SeparateChainingHashMap<>();
    freqs.put('a', 5);
    freqs.put('b', 2);
    freqs.put('c', 1);
    freqs.put('d', 1);
    freqs.put('r', 2);
    freqs.put('!', 1);
    HuffTree tree = buildHuffTree(freqs);
    System.out.println(tree);
}
}

